pkg/upgrade/v1alpha2/kubernetes.go

```diff
@@ -0,0 +1,442 @@
+			err,
+			"failed to list upgradeTask in namespace {%s}",
+			k.namespace,
+		)
+	}
+	return k.list(cli, k.namespace, opts)
+}
+
+// Get returns a upgrade task object
+func (k *Kubeclient) Get(
+	name string,
+	opts metav1.GetOptions,
+) (*apis.UpgradeTask, error) {
+	if strings.TrimSpace(name) == "" {
+		return nil, errors.New(
+			"failed to get upgradeTask: missing upgradeTask name",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get upgradeTask {%s} in namespace {%s}",
+			name,
+			k.namespace,
+		)
+	}
+	return k.get(cli, k.namespace, name, opts)
+}
+
+// Create creates a upgradeTask in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to create upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to create upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.create(cli, k.namespace, upgradeTask)
+}
+
+// Delete deletes a upgradeTask instance from the
+// kubecrnetes cluster
+func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error {
+	if strings.TrimSpace(name) == "" {
+		return errors.New("failed to delete upgradeTask: missing upgradeTask name")
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"failed to delete upgradeTask {%s} in namespace {%s}",
+			name,
+			k.namespace,
+		)
+	}
+	return k.del(cli, k.namespace, name, deleteOpts)
+}
+
+// Patch patches the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Patch(
+	name string,
+	pt types.PatchType,
+	data []byte,
+	subresources ...string,
+) (*apis.UpgradeTask, error) {
+	if len(name) == 0 {
+		return nil, errors.New(
+			"failed to patch upgrade task : missing upgradeTask name",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to patch upgradeTask: {%s}", name)
+	}
+	return k.patch(cli, k.namespace, name, pt, data, subresources...)
+}
+
+// Update updates the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Update(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to udpate upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to update upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.update(cli, k.namespace, upgradeTask)
+}
+
+// GetUpgradeDetailedStatuses builds and returns the UpgradeDetailedStatuses
+// object with given parameters
+func GetUpgradeDetailedStatuses(
+	step apis.UpgradeStep,
```

Comment:
can we accept something like 
```go
func BuildUpgradeDetailsStatuses(step apis.UpgradeStep, status apis.Status)
```

Reply:
So we can avoid checks on args

Reply:
Removed above function and added `IsValidStatus` function